distributed/tests/test_spill.py

```diff
@@ -337,6 +338,9 @@ def test_weakref_cache(tmpdir, cls, expect_cached, size):
     # the same id as a deleted one
     id_x = x.id
     del x
+    # Surprisingly, even on CPython this is needed to ensure that the object is garbage
+    # collected, even if there are no obvious circular references going on
+    gc.collect()
```

Comment:
This does concern me a bit. I don't think the GC should be required here. 

I can reproduce this on py3.10. Maybe there is a CPython bug? Below is what objgraph tells me

```python
        import weakref
        ref = weakref.ref(x)
        del x
        import objgraph
        obj = ref()
        if obj:
            objgraph.show_backrefs(obj)
```

cc @graingert in case something rings a bell

![objgraph-kjmpytgy](https://user-images.githubusercontent.com/8629629/161041289-d517c038-07cf-48b7-b2bf-d7b9cb209b38.png)

Reply:
I ran into similar issues in dask-ms a while back and found I could never rely on the `del x`. If it's an issue of "encouraging" the decref/garbage collector, you might try allocating `x` inside an inner function like so:

```python
   # Run this test twice:

    # - x is smaller than target and is evicted by y;
    # - x is individually larger than target and it never touches fast
    def _inner():
      x = cls(size)
      buf["x"] = x
      if size < 100:
          buf["y"] = cls(60)  # spill x
      assert "x" in buf.slow
      # Test that we update the weakref cache on setitem
      assert (buf["x"] is x) == expect_cached
      # Do not use id_x = id(x), as in CPython id's are C memory addresses and are reused
      # by PyMalloc when you descope objects, so a brand new object might end up having
      # the same id as a deleted one
      id_x = x.id
      del x
     return id_x

  id_x = _inner()
```